provider/pkg/provider/suite_test.go

```diff
@@ -0,0 +1,325 @@
+package provider
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"testing"
+
+	pbempty "github.com/golang/protobuf/ptypes/empty"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	gomegatypes "github.com/onsi/gomega/types"
+	"github.com/pulumi/pulumi-kubernetes/provider/v4/pkg/clients"
+	"github.com/pulumi/pulumi/pkg/v3/resource/provider"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/rpcutil"
+	pulumirpc "github.com/pulumi/pulumi/sdk/v3/proto/go"
+	"google.golang.org/grpc"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	kubeversion "k8s.io/apimachinery/pkg/version"
+	"k8s.io/client-go/discovery"
+	discoveryfake "k8s.io/client-go/discovery/fake"
+	dynamicfake "k8s.io/client-go/dynamic/fake"
+	"k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/rest"
+	kubetesting "k8s.io/client-go/testing"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/kubectl/pkg/scheme"
+)
+
+const (
+	testPluginVersion    = "v0.0.0"
+	testPulumiSchema     = "{}"
+	testTerraformMapping = "{}"
+)
+
+var (
+	testServerVersion = kubeversion.Info{Major: "1", Minor: "29"}
+)
+
+// CheckFailure matches a CheckFailure by property and reason.
+func CheckFailure(prop string, reason gomegatypes.GomegaMatcher) gomegatypes.GomegaMatcher {
+	return And(
+		WithTransform(func(failure *pulumirpc.CheckFailure) string {
+			return failure.GetProperty()
+		}, Equal(prop)),
+		WithTransform(func(failure *pulumirpc.CheckFailure) string {
+			return failure.GetReason()
+		}, reason))
+}
+
+// WriteKubeconfigToString converts a clientcmdapi.Config to a string.
+func WriteKubeconfigToString(config *clientcmdapi.Config) string {
+	contents, err := clientcmd.Write(*config)
+	Expect(err).ToNot(HaveOccurred())
+	return string(contents)
+}
+
+// WriteKubeconfigToFile converts a clientcmdapi.Config to a temporary file.
+func WriteKubeconfigToFile(config *clientcmdapi.Config) string {
+	f, _ := os.CreateTemp("", "kubeconfig-")
```

Comment:
Can we add err back here with an expectation (ie. `Expect(err).ToNot(HaveOccurred())`)? This will help surface temp file creation errors if there is one.